File: Ipopt/contrib/JavaInterface/org/coinor/examples/scalable/LuksanVlcek1.java
/** Copyright (C) 2007 VRTech Industrial Technologies - www.vrtech.com.br.
 * Copyright (C) 2007 Tong Kewei, Beihang University, - www.buaa.edu.cn.
 * All Rights Reserved.
 * This code is published under the Eclipse Public License.
 * 
 * $Id$
 */

package org.coinor.examples.scalable;

/** Implementation of Example 5.1 from "Sparse and Parially Separable Test Problems for Unconstrained and Equality Constrained Optimization" by L. Luksan and J. Vlcek.
 *
 * This code is based on an Ipopt example file with same name.
 * 
 * @author Rafael de Pelegrini Soares, Tong Kewei
 */
public class LuksanVlcek1 extends Scalable
{
	/** Constructor.
	 * Here, gl and gu are the bounds for the constraints.
	 * The original formulation is obtained by setting gl and gu to zero.
	 * Using gl lower than gu allows the obtain a problem formulation with inequality constraints.
	 */
	public LuksanVlcek1(String name, double gl, double gu)
	{
		super(name, gl, gu);
	}

	@Override
	public boolean initialize(int n)
	{
		if( n <= 2 )
		{
			System.out.print("N needs to be at least 3.\n");
			return false;
		}

		// The problem described in LuksanVlcek1.hpp has 4 variables, x[0] through x[3]
		this.n = n;

		m = n - 2;

		nnz_jac_g = m * 3;

		nnz_h_lag = n + n-1;

		// use the C style numbering of matrix indices (starting at 0)
		index_style = C_STYLE;

		// none of the variables have bounds
		x_l = new double[n];
		x_u = new double[n];
		for( int i = 0; i < n; ++i )
		{
			x_l[i] = -1e20;
			x_u[i] =  1e20;
		}

		// Set the bounds for the constraints
		g_l = new double[m];
		g_u = new double[m];
		for( int i = 0; i < m; ++i )
		{
			g_l[i] = gl;
			g_u[i] = gu;
		}

		// set the starting point
		x = new double[n];
		for( int i = 0; i < n/2; ++i )
		{
			x[2*i]  = -1.2;
			x[2*i+1] = 1.0;
		}
		if( n % 2 == 1 )
			x[n-1] = -1.2;

		return true;
	}

	protected boolean get_bounds_info(int n, double[] x_l, double[] x_u,
			int m, double[] g_l, double[] g_u)
	{
		// none of the variables have bounds
		for( int i = 0; i < n; ++i )
		{
			x_l[i] = -1e20;
			x_u[i] =  1e20;
		}

		// Set the bounds for the constraints
		for( int i = 0; i < m; ++i )
		{
			g_l[i] = gl;
			g_u[i] = gu;
		}

		return true;
	}

	protected boolean get_starting_point(int n, boolean init_x, double[] x,
			boolean init_z, double[] z_L, double[] z_U,
			int m, boolean init_lambda,double[] lambda)
	{
		for( int i = 0; i < n/2; ++i )
		{
			x[2*i]  = -1.2;
			x[2*i+1] = 1.0;
		}
		if( n % 2 == 1 )
			x[n-1] = -1.2;

		return true;
	}

	@Override
	protected boolean eval_f(int n, double[] x, boolean new_x, double[] obj_value)
	{
		obj_value[0] = 0.0;
		for( int i = 0; i < n-1; ++i )
		{
			double a1 = x[i] * x[i] - x[i+1];
			double a2 = x[i] - 1.0;
			obj_value[0] += 100.0 * a1 * a1 + a2 * a2;
		}

		return true;
	}

	@Override
	protected boolean eval_g(int n, double[] x, boolean new_x, int m, double[] g)
	{
		for( int i = 0; i < n-2; ++i )
			g[i] = 3.0 * Math.pow(x[i+1], 3.0) + 2.0 * x[i+2] - 5.0 + Math.sin(x[i+1]-x[i+2]) * Math.sin(x[i+1]+x[i+2])
					+ 4.0 * x[i+1] - x[i] * Math.exp(x[i] - x[i+1]) - 3.0;

		return true;
	}

	@Override
	protected boolean eval_grad_f(int n, double[] x, boolean new_x, double[] grad_f)
	{
		grad_f[0] = 0.0;
		for( int i = 0; i < n-1; ++i )
		{
			grad_f[i]  += 400.0 * x[i] * (x[i] * x[i] - x[i+1]) + 2.0 * (x[i] - 1.0);
			grad_f[i+1] = -200.0 * (x[i] * x[i] - x[i+1]);
		}

		return true;
	}

	@Override
	protected boolean eval_jac_g(int n, double[] x, boolean new_x, int m,
			int nele_jac, int[] iRow, int[] jCol, double[] values)
	{
		if( values == null )
		{
			// return the structure of the jacobian
			int ijac=0;
			for( int i = 0; i < n-2; ++i )
			{
				iRow[ijac] = i;
				jCol[ijac] = i;
				ijac++;
				iRow[ijac] = i;
				jCol[ijac] = i+1;
				ijac++;
				iRow[ijac] = i;
				jCol[ijac] = i+2;
				ijac++;
			}
		}
		else
		{
			// return the values of the jacobian of the constraints
			int ijac=0;
			
			for( int i = 0; i < n-2; ++i )
			{
				// x[i]
				values[ijac] = -(1.0 + x[i]) * Math.exp(x[i] - x[i+1]);
				ijac++;
				// x[i+1]
				values[ijac] = 9.0 * x[i+1] * x[i+1]
						+ Math.cos(x[i+1] - x[i+2]) * Math.sin(x[i+1] + x[i+2])
						+ Math.sin(x[i+1] - x[i+2]) * Math.cos(x[i+1] + x[i+2])
						+ 4.0 + x[i] * Math.exp(x[i] - x[i+1]);
				ijac++;
				// x[i+2]
				values[ijac] = 2.0
						- Math.cos(x[i+1] - x[i+2]) * Math.sin(x[i+1] + x[i+2])
						+ Math.sin(x[i+1] - x[i+2]) * Math.cos(x[i+1] + x[i+2]);
				ijac++;
			}
		}

		return true;
	}

	@Override
	protected boolean eval_h(int n, double[] x, boolean new_x,
			double obj_factor, int m, double[] lambda, boolean new_lambda,
			int nele_hess, int[] iRow, int[] jCol, double[] values)
	{
		if( values == null)
		{
			int ihes = 0;
			for( int i = 0; i < n; ++i )
			{
				iRow[ihes] = i;
				jCol[ihes] = i;
				++ihes;
				if( i < n-1 )
				{
					iRow[ihes] = i;
					jCol[ihes] = i+1;
					ihes++;
				}
			}
			assert ihes == nele_hess;
		}
		else
		{
			int ihes = 0;
			for( int i = 0; i < n; ++i )
			{
				// x[i],x[i]
				if( i < n-1 )
				{
					values[ihes] = obj_factor * (2.0 + 400.0 * (3.0 * x[i] * x[i] - x[i+1]));
					if( i < n-2 )
						values[ihes] -= lambda[i] * (2.0 + x[i]) * Math.exp(x[i] - x[i+1]);
				}
				else
					values[ihes] = 0.;

				if( i > 0 )
				{
					// x[i+1]x[i+1]
					values[ihes] += obj_factor * 200.0;
					if( i < n-1 )
						values[ihes] += lambda[i-1]* (18.0 * x[i]
								- 2.0 * Math.sin(x[i] - x[i+1]) * Math.sin(x[i] + x[i+1])
								+ 2.0 * Math.cos(x[i] - x[i+1]) * Math.cos(x[i] + x[i+1])
								- x[i-1] * Math.exp(x[i-1] - x[i]));
				}
				if( i > 1 )
					// x[i+2]x[i+2]
					values[ihes] += lambda[i-2] * (-2.0 * Math.sin(x[i-1] - x[i]) * Math.sin(x[i-1] + x[i])
									- 2.0 * Math.cos(x[i-1] - x[i]) * Math.cos(x[i-1] + x[i]));
				ihes++;

				if( i < n-1 )
				{
					// x[i],x[i+1]
					values[ihes] = obj_factor * (-400.0 * x[i]);
					if( i < n-2 )
						values[ihes] += lambda[i]*(1.+x[i])*Math.exp(x[i]-x[i+1]);
					/*
			        if (i>0) {
			        // x[i+1],x[i+2]
			        values[ihes] +=
			        lambda[i-1]*(  sin(x[i]-x[i+1])*sin(x[i]+x[i+1])
			           + cos(x[i]-x[i+1])*cos(x[i]+x[i+1])
			           - cos(x[i]-x[i+1])*cos(x[i]+x[i+1])
			           - sin(x[i]-x[i+1])*sin(x[i]+x[i+1])
			        );
			        }
					 */
					ihes++;
				}
			}
			assert ihes == nele_hess;
		}

		return true;
	}
}

File: Ipopt/examples/ScalableProblems/LuksanVlcek1.cpp
// Copyright (C) 2005, 2006 International Business Machines and others.
// All Rights Reserved.
// This code is published under the Eclipse Public License.
//
// $Id$
//
// Authors:  Andreas Waechter              IBM    2005-10-127

#include "LuksanVlcek1.hpp"

#ifdef HAVE_CONFIG_H
#include "config.h"
#else
#include "configall_system.h"
#endif

#ifdef HAVE_CMATH
# include <cmath>
#else
# ifdef HAVE_MATH_H
#  include <math.h>
# else
#  error "don't have header file for math"
# endif
#endif

#ifdef HAVE_CSTDIO
# include <cstdio>
#else
# ifdef HAVE_STDIO_H
#  include <stdio.h>
# else
#  error "don't have header file for stdio"
# endif
#endif

using namespace Ipopt;

LuksanVlcek1::LuksanVlcek1(Number g_l, Number g_u)
{
  g_l_ = g_l;
  g_u_ = g_u;
}

bool LuksanVlcek1::InitializeProblem(Index N)
{
  N_=N;
  if (N_<=2) {
    printf("N needs to be at least 3.\n");
    return false;
  }
  return true;
}

// returns the size of the problem
bool LuksanVlcek1::get_nlp_info(Index& n, Index& m, Index& nnz_jac_g,
                                Index& nnz_h_lag, IndexStyleEnum& index_style)
{
  // The problem described in LuksanVlcek1.hpp has 4 variables, x[0] through x[3]
  n = N_;

  m = N_-2;

  nnz_jac_g = m*3;

  nnz_h_lag = n + n-1;

  // use the C style numbering of matrix indices (starting at 0)
  index_style = TNLP::C_STYLE;

  return true;
}

// returns the variable bounds
bool LuksanVlcek1::get_bounds_info(Index n, Number* x_l, Number* x_u,
                                   Index m, Number* g_l, Number* g_u)
{
  // none of the variables have bounds
  for (Index i=0; i<n; i++) {
    x_l[i] = -1e20;
    x_u[i] =  1e20;
  }

  // Set the bounds for the constraints
  for (Index i=0; i<m; i++) {
    g_l[i] = g_l_;
    g_u[i] = g_u_;
  }

  return true;
}

// returns the initial point for the problem
bool LuksanVlcek1::get_starting_point(Index n, bool init_x, Number* x,
                                      bool init_z, Number* z_L, Number* z_U,
                                      Index m, bool init_lambda,
                                      Number* lambda)
{
  if (!init_x || init_z || init_lambda) {
    return false;
  }

  // set the starting point
  for (Index i=0; i<n/2; i++) {
    x[2*i] = -1.2;
    x[2*i+1] = 1.;
  }
  if (n != 2*(n/2)) {
    x[n-1] = -1.2;
  }

  /*
  // DELETEME
  for (Index i=0; i<n; i++) {
    x[i] += 0.001*((Number) i);
  }
  */

  return true;
}

// returns the value of the objective function
bool LuksanVlcek1::eval_f(Index n, const Number* x, bool new_x, Number& obj_value)
{
  obj_value = 0.;
  for (Index i=0; i<N_-1; i++) {
    Number a1 = x[i]*x[i]-x[i+1];
    Number a2 = x[i] - 1.;
    obj_value += 100.*a1*a1 + a2*a2;
  }

  return true;
}

// return the gradient of the objective function grad_{x} f(x)
bool LuksanVlcek1::eval_grad_f(Index n, const Number* x, bool new_x, Number* grad_f)
{
  grad_f[0] = 0.;
  for (Index i=0; i<N_-1; i++) {
    grad_f[i] += 400.*x[i]*(x[i]*x[i]-x[i+1]) + 2.*(x[i]-1.);
    grad_f[i+1] = -200.*(x[i]*x[i]-x[i+1]);
  }

  return true;
}

// return the value of the constraints: g(x)
bool LuksanVlcek1::eval_g(Index n, const Number* x, bool new_x, Index m, Number* g)
{
  for (Index i=0; i<N_-2; i++) {
    g[i] = 3.*pow(x[i+1],3.) + 2.*x[i+2] - 5.
           + sin(x[i+1]-x[i+2])*sin(x[i+1]+x[i+2]) + 4.*x[i+1]
           - x[i]*exp(x[i]-x[i+1]) - 3.;
  }

  return true;
}

// return the structure or values of the jacobian
bool LuksanVlcek1::eval_jac_g(Index n, const Number* x, bool new_x,
                              Index m, Index nele_jac, Index* iRow, Index *jCol,
                              Number* values)
{
  if (values == NULL) {
    // return the structure of the jacobian

    Index ijac=0;
    for (Index i=0; i<N_-2; i++) {
      iRow[ijac] = i;
      jCol[ijac] = i;
      ijac++;
      iRow[ijac] = i;
      jCol[ijac] = i+1;
      ijac++;
      iRow[ijac] = i;
      jCol[ijac] = i+2;
      ijac++;
    }
  }
  else {
    // return the values of the jacobian of the constraints

    Index ijac=0;
    for (Index i=0; i<N_-2; i++) {
      // x[i]
      values[ijac] = -(1.+x[i])*exp(x[i]-x[i+1]);
      ijac++;
      // x[i+1]
      values[ijac] = 9.*x[i+1]*x[i+1]
                     + cos(x[i+1]-x[i+2])*sin(x[i+1]+x[i+2])
                     + sin(x[i+1]-x[i+2])*cos(x[i+1]+x[i+2])
                     + 4. + x[i]*exp(x[i]-x[i+1]);
      ijac++;
      // x[i+2]
      values[ijac] = 2.
                     - cos(x[i+1]-x[i+2])*sin(x[i+1]+x[i+2])
                     + sin(x[i+1]-x[i+2])*cos(x[i+1]+x[i+2]);
      ijac++;
    }
  }

  return true;
}

//return the structure or values of the hessian
bool LuksanVlcek1::eval_h(Index n, const Number* x, bool new_x,
                          Number obj_factor, Index m, const Number* lambda,
                          bool new_lambda, Index nele_hess, Index* iRow,
                          Index* jCol, Number* values)
{
  if (values == NULL) {
    Index ihes=0;
    for (Index i=0; i<N_; i++) {
      iRow[ihes] = i;
      jCol[ihes] = i;
      ihes++;
      if (i<N_-1) {
        iRow[ihes] = i;
        jCol[ihes] = i+1;
        ihes++;
      }
    }
    DBG_ASSERT(ihes == nele_hess);
  }
  else {
    Index ihes=0;
    for (Index i=0; i<N_; i++) {
      // x[i],x[i]
      if (i<N_-1) {
        values[ihes] = obj_factor*(2.+400.*(3.*x[i]*x[i]-x[i+1]));
        if (i<N_-2) {
          values[ihes] -= lambda[i]*(2.+x[i])*exp(x[i]-x[i+1]);
        }
      }
      else {
        values[ihes] = 0.;
      }
      if (i>0) {
        // x[i+1]x[i+1]
        values[ihes] += obj_factor*200.;
        if (i<N_-1) {
          values[ihes] += lambda[i-1]*(18.*x[i]
                                       - 2.*sin(x[i]-x[i+1])*sin(x[i]+x[i+1])
                                       + 2.*cos(x[i]-x[i+1])*cos(x[i]+x[i+1])
                                       - x[i-1]*exp(x[i-1]-x[i]));
        }
      }
      if (i>1) {
        // x[i+2]x[i+2]
        values[ihes] +=
          lambda[i-2]*(- 2.*sin(x[i-1]-x[i])*sin(x[i-1]+x[i])
                       - 2.*cos(x[i-1]-x[i])*cos(x[i-1]+x[i]));
      }
      ihes++;

      if (i<N_-1) {
        // x[i],x[i+1]
        values[ihes] = obj_factor*(-400.*x[i]);
        if (i<N_-2) {
          values[ihes] += lambda[i]*(1.+x[i])*exp(x[i]-x[i+1]);
        }
        /*
        if (i>0) {
        // x[i+1],x[i+2]
        values[ihes] +=
        lambda[i-1]*(  sin(x[i]-x[i+1])*sin(x[i]+x[i+1])
           + cos(x[i]-x[i+1])*cos(x[i]+x[i+1])
           - cos(x[i]-x[i+1])*cos(x[i]+x[i+1])
           - sin(x[i]-x[i+1])*sin(x[i]+x[i+1])
        );
        }
        */
        ihes++;
      }
    }
    DBG_ASSERT(ihes == nele_hess);
  }

  return true;
}

void LuksanVlcek1::finalize_solution(SolverReturn status,
                                     Index n, const Number* x, const Number* z_L, const Number* z_U,
                                     Index m, const Number* g, const Number* lambda,
                                     Number obj_value,
				     const IpoptData* ip_data,
				     IpoptCalculatedQuantities* ip_cq)
{}

